birdhouse/components/weaver/post-docker-compose-up

```diff
@@ -0,0 +1,149 @@
+        msg="Failed to register all provider specified in WEAVER_WPS_PROVIDERS."
+        printf "%s" "${ERROR}Timeout (${WEAVER_WPS_PROVIDERS_MAX_TIME}s)! ${msg}"
+        exit 1
+    fi
+done
+if [ -z "${cookie}" ]; then
+    echo "${ERROR}Failed to retrieve authentication token from Magpie for Weaver WPS providers registration."
+    exit 2
+fi
+
+# parse providers
+WEAVER_WPS_PROVIDERS="$(echo "${WEAVER_WPS_PROVIDERS}" | tr ',' ' ')"
+
+# move on to actual registration of WPS providers
+echo "${PREFIX}Using URL: [${WEAVER_URL}]"
+start_time="$(date -u +%s)"
+ret=1
+for prov in ${WEAVER_WPS_PROVIDERS}; do
+    if [ -z "${prov}" ]; then
+        continue
+    fi
+    prov_url="https://${PAVICS_FQDN_PUBLIC}${TWITCHER_PROTECTED_PATH}/${prov}"
+    prov_cap="${prov_url}?service=WPS&request=GetCapabilities"
+
+    # wait for WPS provider to respond
+    printf "%s" "${PREFIX}Wait for response from remote WPS provider [${prov}] on [${prov_url}]."
+    while true; do
+        # request the URL and obtain the body+http code, then split them for verification
+        resp=$(curl -k -s -m 2 -w "\n%{http_code}" -b "${cookie}" "${prov_cap}")
+        ret=$?   # in case not up yet
+        code=$(echo "${resp}" | tail -n -1)
+        body=$(echo "${resp}" | head -n -1)
+        caps=$(echo "${body}" | grep -c "wps:Capabilities")
+        if [ ${ret} -eq 0 ] && [ "${code}" -eq 200 ] && [ "${caps}" -ne 0 ]; then
+            printf "\n%s\n" "${PREFIX}Got valid response from remote WPS provider [${prov}]."
+            break;
+        fi
+        sleep 1
+        printf "."
+
+        # interrupt if max time reached
+        next_time=$(date -u +%s)
+        delta_time=$(( next_time - start_time ))
+        if [ ${delta_time} -ge "${WEAVER_WPS_PROVIDERS_MAX_TIME}" ]; then
+            msg="Failed to register all provider specified in WEAVER_WPS_PROVIDERS."
+            printf "\n%s\n" "${ERROR}Timeout (${WEAVER_WPS_PROVIDERS_MAX_TIME}s)! ${msg}"
+            exit 3
+        fi
+    done
+
+    # unregister in case of multiple up/down to regenerate from scratch, don't care if NotFound returned
+    echo "${PREFIX}Unregistering any remote WPS provider matching [${prov}]."
+    curl -k -s -m 2 -w "${PREFIX}Delete [${prov}]: %{http_code}" -o /dev/null \
+         -b "${cookie}" \
+         -X DELETE \
+         "${WEAVER_URL}/providers/${prov}"
+
+    # register the new provider and validate
+    printf "\n%s" "${PREFIX}Registering remote WPS provider [${prov}] on [${prov_url}]... "
+    code=$( \
+        curl -k -s -m 2 -w "%{http_code}" -o /dev/null \
+             -b "${cookie}" \
+             -H "Content-Type: application/json" \
+             -X POST \
+             -d "{\"id\": \"${prov}\", \"url\": \"${prov_url}\"}" \
+             "${WEAVER_URL}/providers" \
+    )
+    ret=$?
+    if [ ${ret} -ne 0 ] || [ "${code}" -ne 201 ]; then
+        printf "\n%s\n" "${ERROR}Failed registration of remote WPS provider [${prov} on [${prov_url}]."
+        exit 4
```

Comment:
So as soon as one provider fail, we do not want to try the remaining?

Reply:
There is no point since we unregister and re-register to ensure updated configs are applied.

Reply:
I understand the code this way:
```
for_each provider in [finch, raven, flyingpigeon, ...] :
  wait_for provider
  register provider
  exit completely the loop and drop remaining providers if the above steps for 1 provider take too long
```
So let's say raven take too long, it'll drop the rest and only register finch?

Reply:
Since it exit with non-zero, I would expect all the `docker-compose up` to fail. 
Fixing any relevant config based on logs. 
Then, re-run `docker-compose up`. 
Between `wait_for provider` and `register provider`, there is a `delete provider`. 
Previously deployed `finch` would be re-deployed anyway, hence no need to waste time deploying more of them on previous run if any fails.

Reply:
> Since it exit with non-zero, I would expect all the `docker-compose up` to fail.

FYI, this is a `post-compose` script so `docker-compose up` already executed.  This script do not fail the `docker-compose up` call.

Did you meant fail the `pavics-compose.sh up` script?  I just checked the code and that is not the case either https://github.com/bird-house/birdhouse-deploy/blob/21d98fe5960b820adf768e0d7ca14c21fca2a059/birdhouse/pavics-compose.sh#L199-L205

Each post-compose script should be error resilient and handle any errors themselves.  It should not block other post-compose from other components to run after it.

In fact the design is that the last component has the highest priority to undo or override other components if needed.  The last component is suggested to be the org private config which has the veto right on everything. https://github.com/bird-house/birdhouse-deploy/blob/21d98fe5960b820adf768e0d7ca14c21fca2a059/birdhouse/env.local.example#L59-L71

Reply:
> Did you meant fail the pavics-compose.sh up script? I just checked the code and that is not the case either 

Yes, I meant the whole procedure by `pavics-compose.sh up`. 

I think it is reasonable that any error along the way including pre/post-scripts should fail the whole execution. 

I guess that can be considered a design choice, but I personally prefer early-errors rather than everything "seemingly" working just to get an error much later during unrelated tests/usages. I find the latter makes it much harder to dig up the real cause and fix it, which would have been flagged naturally but error-ing out for known failing cases.